relay-server/src/actors/project_buffer.rs

```diff
@@ -113,66 +159,237 @@ impl FromMessage<RemoveMany> for Buffer {
     }
 }
 
-/// In-memory implementation of the [`Buffer`] interface.
+/// [`Buffer`] interface implementation backed by SQLite.
 #[derive(Debug)]
 pub struct BufferService {
-    /// Contains the cache of the incoming envelopes.
     buffer: BTreeMap<QueueKey, Vec<ManagedEnvelope>>,
+    buffer_guard: Arc<BufferGuard>,
+    config: Option<PersistentBuffer>,
+    // this is half of the `envelope_buffer_size` defined in the [`relay_config::Config`].
+    memory_limit: Option<usize>,
+    count_mem_envelopes: i64,
+    db: Option<Pool<Sqlite>>,
 }
 
 impl BufferService {
-    /// Creates a new [`BufferService`].
-    pub fn new() -> Self {
-        Self {
+    /// Creates a new [`BufferService`] from the provided path to the SQLite database file.
+    pub async fn create(
+        buffer_guard: Arc<BufferGuard>,
+        config: Arc<Config>,
+    ) -> Result<Self, BufferError> {
+        let mut service = Self {
             buffer: BTreeMap::new(),
+            buffer_guard,
+            config: None,
+            memory_limit: None,
+            count_mem_envelopes: 0,
+            db: None,
+        };
+
+        // Only iof persistent buffer enabled, we create the pool and set the config.
+        if let Some(buffer_config) = config.cache_persistent_buffer() {
+            let path = buffer_config.buffer_path().to_owned();
+            let options = SqliteConnectOptions::new()
+                .filename(PathBuf::from("sqlite://").join(&path))
+                .journal_mode(SqliteJournalMode::Wal)
+                .create_if_missing(true);
+
+            let db = SqlitePoolOptions::new()
+                .max_connections(buffer_config.max_connections())
+                .min_connections(buffer_config.min_connections())
+                .connect_with(options)
+                .await?;
+
+            // Set the buffer memory limit, which must not be bigger then `envelope_buffer_size`.
+            let limit = buffer_config
+                .memory_limit()
+                .unwrap_or(config.envelope_buffer_size() / 2);
```

Comment:
I will have to move the `memory_limit()` function out of the `PersistentBuffer` config, since I do not have access to `envelope_buffer_size` in there. That's why I did it here. 
But I might have to move all this getters to top level instead to keep it consistent.